app/models/user.rb

```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  before_save{self.email = email.downcase}
+  validates :name, presence: true, length: {maximum: Settings.user.name_max_length}
+  validates :email, presence: true, length: {maximum: Settings.user.email_max_length},
+                    format: {with: Regexp.new(Settings.user.VALID_EMAIL_REGEX, true)},
+                    uniqueness: true
+  has_secure_password
```

Comment:
hàm này có tác dụng gì em

Reply:
Dạ hàm validates thì để mình cài 1 validation rules cho email để khi user nhập vào thì mình kiểm tra nó hợp lý không ạ.
Còn hàm has_secure_password thì em xin phép được trích lại từ trong sách ra ạ. 

> When included in a model as above, this one method adds the following functionality:
> 
> • The ability to save a securely hashed password_digest attribute to the
> database
> • A pair of virtual attributes19 (password and password_confirmation), including presence validations upon object creation and a validation requiring that they match
> • An authenticate method that returns the user when the password is
> correct (and false otherwise)

Reply:
nói theo ý hiểu của mình đi xem em nắm đc đến đâu
trích như này chú  bảo anh/chị tự hiểu đi hả =))

Reply:
Dạ em xin lỗi chị 😅 Lúc đầu em sợ mình nói sẽ vấp nên mới trích nguyên vậy ạ.
Em xin phép nói theo ý hiểu của em ạ.
has_secure_password tạo cho model users 2 miền password và password_confirmation, 1 miền password_digest để lưu vào database cùng với hàm để chuyển tự động chuyển password về password_digest và cuối cùng hàm authenticate để xác thực users ạ.